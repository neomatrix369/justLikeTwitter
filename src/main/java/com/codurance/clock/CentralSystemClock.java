package com.codurance.clock;

import java.util.Date;

public class CentralSystemClock {

    public Date getCurrentDateTime() {
        return new Date();
    }
}
