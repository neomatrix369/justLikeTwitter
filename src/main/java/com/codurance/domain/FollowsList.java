package com.codurance.domain;

import java.util.HashMap;
import java.util.Map;

public class FollowsList {

    private static final Users USER_DOES_NOT_FOLLOW_ANYONE = null;

    private final Map<User, Users> usersFollowedMapping = new HashMap<>();

    public void addNewUserFollowed(User user, User newFollow) {
        Users existingFollowsList = getFollowsListFor(user);
        existingFollowsList.add(newFollow);
        usersFollowedMapping.put(user, existingFollowsList);
    }

    public Users getFollowsListFor(User user) {
        Users followsList = usersFollowedMapping.get(user);
        if (followsList == USER_DOES_NOT_FOLLOW_ANYONE) {
            return new Users();
        }
        return followsList;
    }
}