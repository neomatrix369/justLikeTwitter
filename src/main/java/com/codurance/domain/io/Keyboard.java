package com.codurance.domain.io;

import java.io.InputStream;
import java.util.Scanner;

import static com.codurance.helper.ImplHelper.STRING_ENCODING;

public class Keyboard implements UserInput {
    private final InputStream inputStream = System.in;

    @Override
    public Scanner getScanner() {
        return new Scanner(inputStream, STRING_ENCODING);
    }
}
