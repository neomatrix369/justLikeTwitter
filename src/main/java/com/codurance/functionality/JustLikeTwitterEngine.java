package com.codurance.functionality;

import com.codurance.clock.CentralSystemClock;
import com.codurance.domain.FollowsList;
import com.codurance.command.UserTypedCommand;
import com.codurance.domain.message.MessageStore;
import com.codurance.functionality.command.CommandExecutor;
import com.codurance.functionality.command.CommandExecutorSelector;

public class JustLikeTwitterEngine {
    private final MessageStore messageStore;
    private final FollowsList followsList;
    private final CentralSystemClock centralSystemClock;

    public JustLikeTwitterEngine(MessageStore messageStore,
                                 FollowsList followsList,
                                 CentralSystemClock centralSystemClock) {
        this.messageStore = messageStore;
        this.followsList = followsList;
        this.centralSystemClock = centralSystemClock;
    }

    public String executeCommand(UserTypedCommand userTypedCommand) {
        CommandExecutorSelector commandExecutorSelector = new CommandExecutorSelector();

        CommandExecutor commandExecutor = commandExecutorSelector.getCommandExecutorFor(userTypedCommand);
        commandExecutor.setMessageStore(messageStore);
        commandExecutor.setCentralSystemClock(centralSystemClock);
        commandExecutor.setFollowsList(followsList);

        return commandExecutor.execute(userTypedCommand);
    }
}