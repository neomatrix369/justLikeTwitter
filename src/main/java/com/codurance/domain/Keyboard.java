package com.codurance.domain;

import java.io.InputStream;
import java.util.Scanner;

import static com.codurance.helper.ImplHelper.STRING_ENCODING;

public class Keyboard {
    private final InputStream inputStream;

    public Keyboard(InputStream inputStream) {
        this.inputStream = inputStream;
    }

    public Scanner getScanner() {
        return new Scanner(inputStream, STRING_ENCODING);
    }
}
