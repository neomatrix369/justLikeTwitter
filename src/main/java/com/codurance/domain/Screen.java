package com.codurance.domain;

import java.io.IOException;
import java.io.OutputStream;

public class Screen implements UserOutput {
    private final OutputStream outputStream = System.out;

    @Override
    public void display(String output) throws IOException {
        outputStream.write(output.getBytes());
    }
}
