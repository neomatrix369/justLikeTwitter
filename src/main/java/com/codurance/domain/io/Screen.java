package com.codurance.domain.io;

import java.io.IOException;
import java.io.PrintStream;

public class Screen implements UserOutput {
    private final PrintStream outputStream = System.out;

    @Override
    public void display(String output) throws IOException {
        outputStream.println(output);
    }
}
