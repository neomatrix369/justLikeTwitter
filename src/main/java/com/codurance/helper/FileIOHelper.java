package com.codurance.helper;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URL;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import static com.codurance.helper.ImplHelper.COLUMN_SEPARATOR;
import static com.codurance.helper.ImplHelper.DATE_COL_INDEX;
import static com.codurance.helper.ImplHelper.MESSAGE_COL_INDEX;
import static com.codurance.helper.ImplHelper.convertToDateFrom;

public class FileIOHelper {

    private FileIOHelper() {
    }

    public static List<Date> loadDatesFrom(Class<?> aClass, String datesForInputFile) throws ParseException, IOException {
        List<Date> result = new ArrayList<>();
        List<String> lines = getTheContentOf(getPathFor(aClass, datesForInputFile).toString());
        for (String eachLine: lines) {
            String[] eachLineSplit = eachLine.split(COLUMN_SEPARATOR);
            String eachDate = eachLineSplit[DATE_COL_INDEX];
            result.add(convertToDateFrom(eachDate));
        }
        return result;
    }

    public static String convertListToStringWithLinefeed(List<String> fileContent) {
        StringBuilder stringBuilder = new StringBuilder();
        for (String eachLine: fileContent) {
            stringBuilder
                    .append(eachLine)
                    .append(System.lineSeparator());
        }

        return stringBuilder.toString();
    }

    public static List<String> getTheContentOf(String fileName) throws IOException {
        File file = new File(fileName);
        Path filePath = file.toPath();
        return Files.readAllLines(filePath, Charset.defaultCharset());
    }

    public static Path getPathFor(Class<?> aClass, String inputFileName) {
        URL fileUrl = aClass.getResource(inputFileName);
        return Paths.get(fileUrl.getPath());
    }

    public static int getNumberOfCommandsIn(Class<?> aClass, String inputFileName) throws IOException {
        Path filePath = getPathFor(aClass, inputFileName);
        List<String> lines = Files.readAllLines(filePath, Charset.defaultCharset());
        return lines.size();
    }

    public static InputStream getFileToReadFrom(Class<?> aClass, String fileName) throws IOException {
        List<String> lines = getTheContentOf(getPathFor(aClass, fileName).toString());

        StringBuilder inputString = new StringBuilder();
        for (String eachLine: lines) {
            String[] splitEachLine = eachLine.split(COLUMN_SEPARATOR);
            inputString
                    .append(splitEachLine[MESSAGE_COL_INDEX])
                    .append(System.lineSeparator());
        }

        return new ByteArrayInputStream(inputString.toString().getBytes());
    }

    public static OutputStream getFileToWriteTo(String fileName) throws FileNotFoundException {
        return new FileOutputStream(fileName);
    }
}
