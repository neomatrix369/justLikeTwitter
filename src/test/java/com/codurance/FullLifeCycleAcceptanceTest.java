package com.codurance;

import com.codurance.clock.CentralSystemClock;
import com.codurance.domain.Keyboard;
import com.codurance.domain.Screen;
import com.codurance.domain.message.MessageStore;
import com.codurance.functionality.JustLikeTwitterEngine;
import com.codurance.helper.FileIOHelper;
import com.codurance.helper.ImplHelper;
import com.codurance.helper.TestHelper;
import com.codurance.domain.FollowsList;
import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.Before;
import org.junit.Test;
import com.codurance.userinterface.IOConsole;

import java.io.IOException;
import java.util.Date;
import java.util.List;

import static org.mockito.AdditionalAnswers.returnsElementsOf;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@UseReporter(DiffReporter.class)
public class FullLifeCycleAcceptanceTest {

    private JustLikeTwitter justLikeTwitter;
    private final CentralSystemClock centralSystemClockMock = mock(CentralSystemClock.class);

    @Before
    public void setUp() throws IOException {
        MessageStore messageStore = new MessageStore();
        FollowsList followsList = new FollowsList();

        JustLikeTwitterEngine justLikeTwitterEngine = new JustLikeTwitterEngine(
                messageStore,
                followsList,
                centralSystemClockMock);

        IOConsole ioConsole = new IOConsole(
                new Keyboard(FileIOHelper.getFileToReadFrom(getClass(), TestHelper.REPLAY_INPUT_FILE)),
                new Screen(FileIOHelper.getFileToWriteTo(TestHelper.ACTUAL_OUTPUT_FILE)),
                ImplHelper.EXTRA_LINEFEED_NEEDED);

        justLikeTwitter = new JustLikeTwitter(ioConsole, justLikeTwitterEngine);
    }

    @Test
    public void givenJustLikeTwitter_whenASeriesOfCommandsArePassedIn_thenResponsesForThemAreSeenInTheConsole() throws Exception {
        // given
        final List<Date> dateTimeForMessages = FileIOHelper.loadDatesFrom(getClass(), TestHelper.REPLAY_INPUT_FILE);
        when(centralSystemClockMock.getCurrentDateTime())
                .thenAnswer(returnsElementsOf(dateTimeForMessages));

        // when
        int numberOfCommands = FileIOHelper.getNumberOfCommandsIn(getClass(), TestHelper.REPLAY_INPUT_FILE);
        justLikeTwitter.run(numberOfCommands);

        // then
        String actualOutputFileContent = FileIOHelper.convertListToStringWithLinefeed(FileIOHelper.getTheContentOf(TestHelper.ACTUAL_OUTPUT_FILE));
        Approvals.verify(actualOutputFileContent);
    }
}