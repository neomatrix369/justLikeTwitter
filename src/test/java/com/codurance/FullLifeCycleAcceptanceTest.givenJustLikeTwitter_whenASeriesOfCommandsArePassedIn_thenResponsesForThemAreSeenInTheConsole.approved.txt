
---------------------------------------------------------------------------------------
Running JustLikeTwitter console app
Use Ctrl-C to exit the program.
Only sunny day scenarios have been covered.
---------------------------------------------------------------------------------------

Command-line help (usage with examples):

  - Posting to personal timeline: <user name> -> <message>
  for e.g.
     > Alice -> I'm having a great time

  - Reading any user's timeline: <user name>
  for e.g.
     > Alice
     I'm having a great time (2 seconds ago)

  - Following another user: <user name> follows <another user>
  for e.g.
     > Alice follows Bob

  - Display user's wall : <user name> wall
  for e.g.
     > Alice wall
     Bob - I'm in New York today! Anyone wants to have a coffee? (2 seconds ago)
     Alice - I'm having a great time (5 minutes ago)

> 

> 

> 

> 
I love the weather today (5 minutes ago)

> 
Good game though. (1 minute ago)
Damn! We lost! (2 minutes ago)

> 

> 

> 
Charlie - I'm in New York today! Anyone wants to have a coffee? (1 second ago)
Alice - I love the weather today (7 minutes ago)

> 

> 
Charlie - I'm in New York today! Anyone wants to have a coffee? (2 minutes ago)
Bob - Good game though. (1 hour ago)
Bob - Damn! We lost! (1 hour ago)
Alice - I love the weather today (1 hour ago)

> 
Alice - I love the weather today (1 hour ago)
